File: regresion_polinomica.py
# -*- coding: utf-8 -*-
"""
Created on Tue Mar 16 16:01:34 2021

@author: hp
"""
#regresion polinómica
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el data set
dataset = pd.read_csv("Position_Salaries.csv")
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

#no se necesita dividir el dataset porque no tendría sentido, son muy pocos.
#no hay que escalar, porque queremos más bien ver la relaciónes no lineales con polinomial

#ajustar la regresion lineal com dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

#ajustar la regresion polinómica con el dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)#qué pasaría si le agrego grado 3 0 4?, se mejora con 3, y se mejora más con grado 4
X_poly = poly_reg.fit_transform(X)#transformación polinómica.
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

#visualización de los resultados del modelo lineal
plt.scatter(X, y, color="red")
plt.plot(X, lin_reg.predict(X), color="blue")#predecir valores para el conjunto de datos x
plt.title("modelo de regresion lineal")
plt.xlabel("posicion del empleado")
plt.ylabel("sueldo en $")
plt.show()
#visualización de los resultados del modelo POLINOMICO
X_grid = np.arange(min(X), max(X), 0.1) #para hacer puntos intermedios y que no se vea como rectas, si no con suavidad
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color="red")
plt.plot(X, lin_reg_2.predict(poly_reg.fit_transform(X)), color="blue")#predecir valores para el conjunto de datos x_poly solo al modelo de predicción.
plt.title("modelo de regresion polinómica")
plt.xlabel("posicion del empleado")
plt.ylabel("sueldo en $")
plt.show()

#prediccion de nuestros modelos
lin_reg.predict(6.5)


lin_reg_2.predict(poly_reg.fit_transform(6.5))#meter un valor para ver cómo está el modelo.

File: Regresion_polinomica.R
# Importar el dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[, 2:3]

# Dividir los datos en conjunto de entrenamiento y conjunto de test
# install.packages("caTools")
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# testing_set = subset(dataset, split == FALSE)

# Escalado de valores
# training_set[,2:3] = scale(training_set[,2:3])
# testing_set[,2:3] = scale(testing_set[,2:3])

# ajustar modelo de regresion lineal con el conjunto de datos
lin_reg = lm(formula = Salary ~ .,
             data = dataset)


# ajustar el modelo de regresión polinómica con el conjunto de datos
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ .,
              data = dataset)
#visualizacion modelo lineal
library(ggplot2)
ggplot() +
  geom_point(aes(x=dataset$Level, y= dataset$Salary),
             color = "red") +
  geom_line(aes(x= dataset$Level,y=predict(lin_reg, newdata = dataset)),
            color = "blue") +
  ggtitle("prediccion lineal del sueldo en funcion del nivel del empleado") +
  xlab("nivel del empleado") +
  ylab("sueldo en $")
#visualización de modelo polinomico
ggplot() +
  geom_point(aes(x=dataset$Level, y= dataset$Salary),
             color = "red") +
  geom_line(aes(x= dataset$Level,y=predict(poly_reg, newdata = dataset)),
            color = "blue") +
  ggtitle("prediccion polinomico del sueldo en funcion del nivel del empleado") +
  xlab("nivel del empleado") +
  ylab("sueldo en $")
#prediccion de nuevos resultados con regresion lineal
y_pred = predict(lin_reg, newdata = data.frame(Level= 6.5))#hay que crear un dataframe en R para una nueva prediccion
#prediccion de nuevos resultados con regresion polinomica
y_pred_poli = predict(poly_reg, newdata = data.frame(Level= 6.5,
                                                     Level2 = 6.5^2,
                                                     Level3 = 6.5^3,
                                                     Level4 = 6.5^4))
